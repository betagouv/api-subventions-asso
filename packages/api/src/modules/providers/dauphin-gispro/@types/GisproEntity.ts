export default interface GisproEntity {
    ej: string;
    codeActionDossier: string;
    codeProjet: string;
    siret: string;
    directionGestionnaire?: string;
    exercise: number;
}
