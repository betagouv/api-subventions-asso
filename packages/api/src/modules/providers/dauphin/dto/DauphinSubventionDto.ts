/**
 * /!\ This DTO is not complete, because no docs are found so we complete as we go along.
 * Pease check if attribute is already here before use this
 */
export default interface DauphinSubventionDto {
    id: string;
    reference: string;
    _document: {
        dateVersion: string;
    };
    history: {
        events: { date: string }[];
        begin: { date: string };
    };
    intituleProjet: string;
    description?: {
        value: string;
    };
    virtualStatusLabel: string;
    dateDemande: string;
    exerciceBudgetaire: number;
    financeursPrivilegies?: [
        {
            title: string;
        }
    ];
    thematique?: {
        title?: string;
    };
    status: string;
    demandeur: {
        SIRET: {
            complet: string;
            SIREN: string;
        };
    };

    planFinancement: [
        {
            current: boolean;
            recette?: {
                postes?: [
                    {
                        sousPostes?: [
                            {
                                lignes?: [
                                    {
                                        dispositifEligible: boolean;
                                        montant: {
                                            ht: number;
                                        };
                                        financement?: {
                                            montantVote?: {
                                                ht: number;
                                            };
                                        };
                                    }
                                ];
                            }
                        ];
                    }
                ];
            };
        }
    ];
}
