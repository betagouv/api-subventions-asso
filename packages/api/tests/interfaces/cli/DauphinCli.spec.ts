import dauphinPort from "../../../src/dataProviders/db/providers/dauphin/dauphin.port";
import gisproPort from "../../../src/dataProviders/db/providers/gispro.port";
import DauphinSubventionDto from "../../../src/modules/providers/dauphin-gispro/dto/DauphinSubventionDto";
import GisproEntity from "../../../src/modules/providers/dauphin-gispro/@types/GisproEntity";
import applicationFlatPort from "../../../src/dataProviders/db/applicationFlat/applicationFlat.port";
import DauphinCli from "../../../src/interfaces/cli/Dauphin.cli";

describe("Dauphin cli", () => {
    // there are other methods but that are not meant to be used a lot anymore since
    // dauphin data is switched to osiris provider

    describe("initApplicationFlat", () => {
        const cli = new DauphinCli();
        const ENTITY1 = {
            exerciceBudgetaire: 2022,
            demandeur: { SIRET: { complet: "01234567890123", SIREN: "012345678" } },
            codeActionProjet: "12345678",
            reference: "12345678",
            intituleProjet: "projet1 - titre",
            periode: "PLURIANNUELLE",
            dateDemande: "2021-12-07",
            thematique: { title: "thematique - titre" },
            virtualStatusLabel: "Justifiée",
            description: { value: "description d'une première action" },
            planFinancement: [
                {
                    current: true,
                    recette: {
                        postes: [
                            {
                                reference: "74",
                                sousPostes: [
                                    {
                                        reference: "74etat",
                                        lignes: [
                                            {
                                                dispositifEligible: true,
                                                montant: { ht: 4200 },
                                                financement: {
                                                    financeur: { typeFinanceur: "FINANCEURPRIVILEGIE", titre: "ville" },
                                                    montantVote: { ht: 4000 },
                                                },
                                            },
                                        ],
                                    },
                                ],
                            },
                        ],
                    },
                },
            ],
            updateDate: new Date("2025-07-07"),
        } as Partial<DauphinSubventionDto>;

        const ENTITY2 = {
            exerciceBudgetaire: 2022,
            demandeur: { SIRET: { complet: "01234567890123", SIREN: "012345678" } },
            codeActionProjet: "87654321",
            reference: "87654321",
            intituleProjet: "projet2 - titre",
            periode: "PLURIANNUELLE",
            dateDemande: "2021-12-07",
            thematique: { title: "thematique - titre" },
            virtualStatusLabel: "Justifiée",
            description: { value: "description d'une première action" },
            planFinancement: [
                {
                    current: true,
                    recette: {
                        postes: [
                            {
                                reference: "74",
                                sousPostes: [
                                    {
                                        reference: "74etat",
                                        lignes: [
                                            {
                                                dispositifEligible: true,
                                                montant: { ht: 4200 },
                                                financement: {
                                                    financeur: { typeFinanceur: "FINANCEURPRIVILEGIE", titre: "ville" },
                                                    montantVote: { ht: 4000 },
                                                },
                                            },
                                        ],
                                    },
                                ],
                            },
                        ],
                    },
                },
            ],
            updateDate: new Date("2025-07-07"),
        } as Partial<DauphinSubventionDto>;
        const GISPRO1 = {
            ej: "ej",
            codeActionDossier: "12345678",
            codeProjet: "projet",
            siret: "01234567890123",
            directionGestionnaire: "gestionnaire",
            exercise: 2022,
        } as GisproEntity;
        const GISPRO2 = {
            ej: "ej",
            codeActionDossier: "87654321",
            codeProjet: "projet",
            siret: "01234567890123",
            directionGestionnaire: "gestionnaire",
            exercise: 2022,
        } as GisproEntity;

        it("saves adapted simple dauphin data", async () => {
            await dauphinPort.upsert({ dauphin: ENTITY1 as DauphinSubventionDto });
            await gisproPort.insertMany([GISPRO1]);
            await cli.initApplicationFlat();
            const actual = await applicationFlatPort.findAll();
            expect(actual).toMatchSnapshot();
        });

        it("groups actions of same application", async () => {
            await dauphinPort.upsert({ dauphin: ENTITY1 as DauphinSubventionDto });
            await dauphinPort.upsert({ dauphin: ENTITY2 as DauphinSubventionDto });
            await gisproPort.insertMany([GISPRO1, GISPRO2]);
            await cli.initApplicationFlat();
            const actual = await applicationFlatPort.findAll();
            expect(actual).toMatchSnapshot();
        });

        it("filters out other allocators than expected", async () => {
            const biggerPlanFinancement = [
                {
                    current: true,
                    recette: {
                        postes: [
                            {
                                reference: "74", // good
                                sousPostes: [
                                    {
                                        reference: "74etat",
                                        lignes: [
                                            {
                                                // good typeFinanceur
                                                dispositifEligible: true,
                                                montant: { ht: 1200 },
                                                financement: {
                                                    financeur: { typeFinanceur: "FINANCEURPRIVILEGIE", titre: "ville" },
                                                    montantVote: { ht: 1000 },
                                                },
                                            },
                                            {
                                                // other good typeFinanceur
                                                dispositifEligible: true,
                                                montant: { ht: 1200 },
                                                financement: {
                                                    financeur: {
                                                        typeFinanceur: "FINANCEURPRIVILEGIE",
                                                        titre: "région",
                                                    },
                                                    montantVote: { ht: 1000 },
                                                },
                                            },
                                            {
                                                // bad typeFinanceur
                                                dispositifEligible: true,
                                                montant: { ht: 1200 },
                                                financement: {
                                                    financeur: { typeFinanceur: "PETITFINANCEUR", titre: "ville" },
                                                    montantVote: { ht: 1000 },
                                                },
                                            },
                                        ],
                                    }, // good
                                    {
                                        reference: "74other", // bad
                                        lignes: [
                                            {
                                                dispositifEligible: true,
                                                montant: { ht: 1200 },
                                                financement: {
                                                    financeur: { typeFinanceur: "FINANCEURPRIVILEGIE", titre: "ville" },
                                                    montantVote: { ht: 1000 },
                                                },
                                            },
                                        ],
                                    },
                                ],
                            },
                            {
                                reference: "75", // bad
                                sousPostes: [
                                    {
                                        reference: "75other",
                                        lignes: [
                                            {
                                                dispositifEligible: true,
                                                montant: { ht: 1200 },
                                                financement: {
                                                    financeur: { typeFinanceur: "FINANCEURPRIVILEGIE", titre: "ville" },
                                                    montantVote: { ht: 1000 },
                                                },
                                            },
                                        ],
                                    },
                                ],
                            },
                        ],
                    },
                },
            ];
            await dauphinPort.upsert({
                dauphin: { ...ENTITY1, planFinancement: biggerPlanFinancement } as DauphinSubventionDto,
            });
            await gisproPort.insertMany([GISPRO1]);
            await cli.initApplicationFlat();
            const actual = await applicationFlatPort.findAll();
            expect(actual).toMatchSnapshot();
        });

        it("keeps action-level dauphin data if nothing is found from gispro", async () => {
            await dauphinPort.upsert({ dauphin: ENTITY1 as DauphinSubventionDto });
            await dauphinPort.upsert({ dauphin: ENTITY2 as DauphinSubventionDto });
            await cli.initApplicationFlat();
            const actual = await applicationFlatPort.findAll();
            expect(actual).toMatchSnapshot();
        });

        it("keeps action-level dauphin data if several records are found from gispro", async () => {
            await dauphinPort.upsert({ dauphin: ENTITY1 as DauphinSubventionDto });
            await dauphinPort.upsert({ dauphin: ENTITY2 as DauphinSubventionDto });
            await gisproPort.insertMany([GISPRO1, { ...GISPRO1, ej: "autreEJ" }]);

            await cli.initApplicationFlat();
            const actual = await applicationFlatPort.findAll();
            expect(actual).toMatchSnapshot();
        });
    });
});
