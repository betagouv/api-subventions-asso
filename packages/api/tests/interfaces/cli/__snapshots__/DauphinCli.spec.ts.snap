// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dauphin cli initApplicationFlat filters out other allocators than expected 1`] = `
[
  {
    "allocatorId": null,
    "allocatorIdType": null,
    "allocatorName": "",
    "applicationId": "dauphin-12345678",
    "applicationProviderId": "12345678",
    "beneficiaryEstablishmentId": "01234567890123",
    "beneficiaryEstablishmentIdType": "siret",
    "budgetaryYear": 2022,
    "cofinancersIdType": null,
    "cofinancersNames": null,
    "cofinancingRequested": null,
    "confinancersId": null,
    "conventionDate": null,
    "decisionDate": null,
    "decisionReference": null,
    "depositDate": 2021-12-07T00:00:00.000Z,
    "ej": "ej",
    "grantedAmount": 2000,
    "idRAE": null,
    "instructiveDepartementId": null,
    "instructiveDepartmentIdType": null,
    "instructiveDepartmentName": "gestionnaire",
    "joinKeyDesc": null,
    "joinKeyId": null,
    "managingAuthorityId": "N/A",
    "managingAuthorityIdType": "N/A",
    "managingAuthorityName": "N/A",
    "nature": "Aide en numéraire",
    "object": "projet1 - titre",
    "paymentCondition": null,
    "paymentConditionDesc": null,
    "paymentId": "01234567890123 - ej - 2022",
    "paymentPeriodDates": null,
    "pluriannual": true,
    "pluriannualYears": null,
    "provider": "dauphin",
    "requestYear": 2021,
    "requestedAmount": 2400,
    "scheme": "thematique - titre - Politique de la ville",
    "statusLabel": "Accordé",
    "subScheme": null,
    "subventionPercentage": null,
    "totalAmount": null,
    "ueNotification": null,
    "uniqueId": "dauphin-12345678-2022",
    "updateDate": 2025-07-07T00:00:00.000Z,
  },
]
`;

exports[`Dauphin cli initApplicationFlat groups actions of same application 1`] = `
[
  {
    "allocatorId": null,
    "allocatorIdType": null,
    "allocatorName": "",
    "applicationId": "dauphin-87654321|12345678",
    "applicationProviderId": "87654321|12345678",
    "beneficiaryEstablishmentId": "01234567890123",
    "beneficiaryEstablishmentIdType": "siret",
    "budgetaryYear": 2022,
    "cofinancersIdType": null,
    "cofinancersNames": null,
    "cofinancingRequested": null,
    "confinancersId": null,
    "conventionDate": null,
    "decisionDate": null,
    "decisionReference": null,
    "depositDate": 2021-12-07T00:00:00.000Z,
    "ej": "ej",
    "grantedAmount": 8000,
    "idRAE": null,
    "instructiveDepartementId": null,
    "instructiveDepartmentIdType": null,
    "instructiveDepartmentName": "gestionnaire",
    "joinKeyDesc": null,
    "joinKeyId": null,
    "managingAuthorityId": "N/A",
    "managingAuthorityIdType": "N/A",
    "managingAuthorityName": "N/A",
    "nature": "Aide en numéraire",
    "object": "projet2 - titre|projet1 - titre",
    "paymentCondition": null,
    "paymentConditionDesc": null,
    "paymentId": "01234567890123 - ej - 2022",
    "paymentPeriodDates": null,
    "pluriannual": true,
    "pluriannualYears": null,
    "provider": "dauphin",
    "requestYear": 2021,
    "requestedAmount": 8400,
    "scheme": "thematique - titre - Politique de la ville",
    "statusLabel": "Accordé",
    "subScheme": null,
    "subventionPercentage": null,
    "totalAmount": null,
    "ueNotification": null,
    "uniqueId": "dauphin-87654321|12345678-2022",
    "updateDate": 2025-07-07T00:00:00.000Z,
  },
]
`;

exports[`Dauphin cli initApplicationFlat keeps action-level dauphin data if nothing is found from gispro 1`] = `
[
  {
    "allocatorId": null,
    "allocatorIdType": null,
    "allocatorName": "",
    "applicationId": "dauphin-87654321",
    "applicationProviderId": "87654321",
    "beneficiaryEstablishmentId": "01234567890123",
    "beneficiaryEstablishmentIdType": "siret",
    "budgetaryYear": 2022,
    "cofinancersIdType": null,
    "cofinancersNames": null,
    "cofinancingRequested": null,
    "confinancersId": null,
    "conventionDate": null,
    "decisionDate": null,
    "decisionReference": null,
    "depositDate": 2021-12-07T00:00:00.000Z,
    "ej": null,
    "grantedAmount": 4000,
    "idRAE": null,
    "instructiveDepartementId": null,
    "instructiveDepartmentIdType": null,
    "instructiveDepartmentName": "",
    "joinKeyDesc": null,
    "joinKeyId": null,
    "managingAuthorityId": "N/A",
    "managingAuthorityIdType": "N/A",
    "managingAuthorityName": "N/A",
    "nature": "Aide en numéraire",
    "object": "projet2 - titre",
    "paymentCondition": null,
    "paymentConditionDesc": null,
    "paymentId": "01234567890123 -  - 2022",
    "paymentPeriodDates": null,
    "pluriannual": true,
    "pluriannualYears": null,
    "provider": "dauphin",
    "requestYear": 2021,
    "requestedAmount": 4200,
    "scheme": "thematique - titre - Politique de la ville",
    "statusLabel": "Accordé",
    "subScheme": null,
    "subventionPercentage": null,
    "totalAmount": null,
    "ueNotification": null,
    "uniqueId": "dauphin-87654321-2022",
    "updateDate": 2025-07-07T00:00:00.000Z,
  },
  {
    "allocatorId": null,
    "allocatorIdType": null,
    "allocatorName": "",
    "applicationId": "dauphin-12345678",
    "applicationProviderId": "12345678",
    "beneficiaryEstablishmentId": "01234567890123",
    "beneficiaryEstablishmentIdType": "siret",
    "budgetaryYear": 2022,
    "cofinancersIdType": null,
    "cofinancersNames": null,
    "cofinancingRequested": null,
    "confinancersId": null,
    "conventionDate": null,
    "decisionDate": null,
    "decisionReference": null,
    "depositDate": 2021-12-07T00:00:00.000Z,
    "ej": null,
    "grantedAmount": 4000,
    "idRAE": null,
    "instructiveDepartementId": null,
    "instructiveDepartmentIdType": null,
    "instructiveDepartmentName": "",
    "joinKeyDesc": null,
    "joinKeyId": null,
    "managingAuthorityId": "N/A",
    "managingAuthorityIdType": "N/A",
    "managingAuthorityName": "N/A",
    "nature": "Aide en numéraire",
    "object": "projet1 - titre",
    "paymentCondition": null,
    "paymentConditionDesc": null,
    "paymentId": "01234567890123 -  - 2022",
    "paymentPeriodDates": null,
    "pluriannual": true,
    "pluriannualYears": null,
    "provider": "dauphin",
    "requestYear": 2021,
    "requestedAmount": 4200,
    "scheme": "thematique - titre - Politique de la ville",
    "statusLabel": "Accordé",
    "subScheme": null,
    "subventionPercentage": null,
    "totalAmount": null,
    "ueNotification": null,
    "uniqueId": "dauphin-12345678-2022",
    "updateDate": 2025-07-07T00:00:00.000Z,
  },
]
`;

exports[`Dauphin cli initApplicationFlat saves adapted simple dauphin data 1`] = `
[
  {
    "allocatorId": null,
    "allocatorIdType": null,
    "allocatorName": "",
    "applicationId": "dauphin-12345678",
    "applicationProviderId": "12345678",
    "beneficiaryEstablishmentId": "01234567890123",
    "beneficiaryEstablishmentIdType": "siret",
    "budgetaryYear": 2022,
    "cofinancersIdType": null,
    "cofinancersNames": null,
    "cofinancingRequested": null,
    "confinancersId": null,
    "conventionDate": null,
    "decisionDate": null,
    "decisionReference": null,
    "depositDate": 2021-12-07T00:00:00.000Z,
    "ej": "ej",
    "grantedAmount": 4000,
    "idRAE": null,
    "instructiveDepartementId": null,
    "instructiveDepartmentIdType": null,
    "instructiveDepartmentName": "gestionnaire",
    "joinKeyDesc": null,
    "joinKeyId": null,
    "managingAuthorityId": "N/A",
    "managingAuthorityIdType": "N/A",
    "managingAuthorityName": "N/A",
    "nature": "Aide en numéraire",
    "object": "projet1 - titre",
    "paymentCondition": null,
    "paymentConditionDesc": null,
    "paymentId": "01234567890123 - ej - 2022",
    "paymentPeriodDates": null,
    "pluriannual": true,
    "pluriannualYears": null,
    "provider": "dauphin",
    "requestYear": 2021,
    "requestedAmount": 4200,
    "scheme": "thematique - titre - Politique de la ville",
    "statusLabel": "Accordé",
    "subScheme": null,
    "subventionPercentage": null,
    "totalAmount": null,
    "ueNotification": null,
    "uniqueId": "dauphin-12345678-2022",
    "updateDate": 2025-07-07T00:00:00.000Z,
  },
]
`;
